### ISSUE TITLE
InconsistentStateException happened when using single silo

### ISSUE THREAD

**Issue by Itoktsnhc** [dotnet/orleans#6128]
I am using direct client with single silo in same process in asp.net core。
Can this InconsistentStateException  occur in this situation?
net core ：3.0.0
Storage Provider：Azure table Persistence.AzureStorage
orleans：3.0.0

**Comment by Itoktsnhc** [comment]
silo:
https://github.com/Itoktsnhc/JobTrackerX.Orleans/blob/master/JobTrackerX.WebApi/Services/Background/InProcessSilo.cs

grain : https://github.com/Itoktsnhc/JobTrackerX.Orleans/blob/master/JobTrackerX.Grains/JobGrain.cs

**Comment by Itoktsnhc** [comment]
> InconsistentStateException: Table storage condition not Satisfied.  GrainType: JobTrackerX.Grains.JobGrain, GrainId: GrainReference:*grn/ECD4041A/13bcb8d2, TableName: JobEntities, StoredETag: Unknown, CurrentETag: W/"datetime'2019-11-17T00%3A34%3A22.375013Z'" Expected Etag=Unknown Received Etag=W/"datetime'2019-11-17T00%3A34%3A22.375013Z'" Microsoft.Azure.Cosmos.Table.StorageException: Precondition Failed
   at Microsoft.Azure.Cosmos.Table.RestExecutor.TableCommand.Executor.ExecuteAsync[T](RESTCommand`1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token)
   at Orleans.Persistence.AzureStorage.AzureTableDataManager`1.UpdateTableEntryAsync(T data, String dataEtag)
   at Orleans.Storage.AzureTableGrainStorage.GrainStateTableDataManager.Write(GrainStateRecord record)
   at Orleans.Storage.AzureTableGrainStorage.DoOptimisticUpdate(Func`1 updateOperation, String grainType, GrainReference grainReference, String tableName, String currentETag)
Request Information
RequestID:df5026fe-3002-0047-5edf-9ccaf8000000
RequestDate:Sun, 17 Nov 2019 00:36:49 GMT
StatusMessage:Precondition Failed
ErrorCode:
ErrorMessage:The update condition specified in the request was not satisfied.
RequestId:df5026fe-3002-0047-5edf-9ccaf8000000
Time:2019-11-17T00:36:49.4481895Z

> Exc level 0: Orleans.Storage.TableStorageUpdateConditionNotSatisfiedException: Table storage condition not Satisfied.  GrainType: JobTrackerX.Grains.JobGrain, GrainId: GrainReference:*grn/ECD4041A/13bcb8d2, TableName: JobEntities, StoredETag: Unknown, CurrentETag: W/"datetime'2019-11-17T00%3A34%3A22.375013Z'"
   at Orleans.Storage.AzureTableGrainStorage.DoOptimisticUpdate(Func`1 updateOperation, String grainType, GrainReference grainReference, String tableName, String currentETag)
   at Orleans.Storage.AzureTableGrainStorage.WriteStateAsync(String grainType, GrainReference grainReference, IGrainState grainState)
   at Orleans.Core.StateStorageBridge`1.WriteStateAsync()
Exc level 1: Microsoft.Azure.Cosmos.Table.StorageException: Precondition Failed
   at Microsoft.Azure.Cosmos.Table.RestExecutor.TableCommand.Executor.ExecuteAsync[T](RESTCommand`1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token)
   at Orleans.Persistence.AzureStorage.AzureTableDataManager`1.UpdateTableEntryAsync(T data, String dataEtag)
   at Orleans.Storage.AzureTableGrainStorage.GrainStateTableDataManager.Write(GrainStateRecord record)
   at Orleans.Storage.AzureTableGrainStorage.DoOptimisticUpdate(Func`1 updateOperation, String grainType, GrainReference grainReference, String tableName, String currentETag)
InconsistentStateException: Table storage condition not Satisfied.  GrainType: JobTrackerX.Grains.JobGrain, GrainId: GrainReference:*grn/ECD4041A/13bcb8d2, TableName: JobEntities, StoredETag: Unknown, CurrentETag: W/"datetime'2019-11-17T00%3A34%3A22.375013Z'" Expected Etag=Unknown Received Etag=W/"datetime'2019-11-17T00%3A34%3A22.375013Z'" Microsoft.Azure.Cosmos.Table.StorageException: Precondition Failed
   at Microsoft.Azure.Cosmos.Table.RestExecutor.TableCommand.Executor.ExecuteAsync[T](RESTCommand`1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token)
   at Orleans.Persistence.AzureStorage.AzureTableDataManager`1.UpdateTableEntryAsync(T data, String dataEtag)
   at Orleans.Storage.AzureTableGrainStorage.GrainStateTableDataManager.Write(GrainStateRecord record)
   at Orleans.Storage.AzureTableGrainStorage.DoOptimisticUpdate(Func`1 updateOperation, String grainType, GrainReference grainReference, String tableName, String currentETag)
Request Information
RequestID:df5026fe-3002-0047-5edf-9ccaf8000000
RequestDate:Sun, 17 Nov 2019 00:36:49 GMT
StatusMessage:Precondition Failed
ErrorCode:
ErrorMessage:The update condition specified in the request was not satisfied.
RequestId:df5026fe-3002-0047-5edf-9ccaf8000000
Time:2019-11-17T00:36:49.4481895Z

> InconsistentStateException: Table storage condition not Satisfied.  GrainType: JobTrackerX.Grains.JobGrain, GrainId: GrainReference:*grn/ECD4041A/13bcb8d2, TableName: JobEntities, StoredETag: Unknown, CurrentETag: W/"datetime'2019-11-17T00%3A34%3A22.375013Z'" Expected Etag=Unknown Received Etag=W/"datetime'2019-11-17T00%3A34%3A22.375013Z'" Microsoft.Azure.Cosmos.Table.StorageException: Precondition Failed
   at Microsoft.Azure.Cosmos.Table.RestExecutor.TableCommand.Executor.ExecuteAsync[T](RESTCommand`1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token)
   at Orleans.Persistence.AzureStorage.AzureTableDataManager`1.UpdateTableEntryAsync(T data, String dataEtag)
   at Orleans.Storage.AzureTableGrainStorage.GrainStateTableDataManager.Write(GrainStateRecord record)
   at Orleans.Storage.AzureTableGrainStorage.DoOptimisticUpdate(Func`1 updateOperation, String grainType, GrainReference grainReference, String tableName, String currentETag)
Request Information
RequestID:df5026fe-3002-0047-5edf-9ccaf8000000
RequestDate:Sun, 17 Nov 2019 00:36:49 GMT
StatusMessage:Precondition Failed
ErrorCode:
ErrorMessage:The update condition specified in the request was not satisfied.
RequestId:df5026fe-3002-0047-5edf-9ccaf8000000
Time:2019-11-17T00:36:49.4481895Z

**Comment by jason-bragg** [comment]
Inconsistent state means only that the state cached in the grain has become out of date with storage.  

This can happen in many ways.  Most common is a timeout on write.  If a write times out, then it's not clear if the state was updated in storage or not.  If it was, then it will be out of date with the state cached in the grain, because the grain state's etag would not have been updated.  This would result in the next write call generating an inconsistent state error.

The safest path, imo, is to always deactivate the grain on storage operation (read, write, or clear) failure (of any kind).  It's a bit heavy handed, and may not be the right pattern if your service is performance critical, but is the cleanest, safest pattern.

**Comment by SebastianStehle** [comment]
I thought that this happens automatically, 

see: https://github.com/dotnet/orleans/blob/37c6c735747db2f626ff66c58be260fc7308d7a0/src/Orleans.Runtime/Core/InsideRuntimeClient.cs#L414

**Comment by jason-bragg** [comment]
On inconsistent state exception the grain is deactivated automatically. Correct.  This means one 'shouldn't' need do anything for the service to self-correct.  However, it does mean the service would need encounter an error for it to self-correct.

Consider the timeout scenario I described.  If a grain call updates state then times out on the write, that call will fail (due to the timeout), but so could the next modification of state, since it could throw an Inconsistent State exception.  After the second error the grain will deactivate and recover on next call.  This means the service will encounter two errors, the timeout, then the inconsistent state.  If the application logic caught the timeout and deactivated the grain at that point, it would only encounter the single error.

Further, since the state was already modified, but the write timed out, the modified state would still be in the grain memory, even though it may not be the same as what is in storage.  Any read calls that operate off the in-memory state may be serving incorrect results, since it comes from the state cached in memory, which may not be the same as storage.  How much of a problem this is depends on the application and how important it is for state to be consistent.

In short, if a storage write times out, but the operation was committed, then the state in the grain is the same as storage, but the etag is out of date, so the next call will trigger an inconsistent state exception, but if the storage operation failed, then the etag would be ok, but the in memory state would be out of date with storage, because the state in storage had not been updated.  In either case, something is out of date which could harm the service.

**Comment by Itoktsnhc** [comment]
@jason-bragg 
Thanks for explaining. I did see access to azure table with a delay of about 50ms when Exception happened
Can I configure the timeout for the  state write operation?
